### ISSUE TITLE
The prediction layer is not learning without batchnorm

### ISSUE THREAD

**Issue by zhongpeixiang** [JasonForJoy/DIM#3]
I implemented your model in PyTorch. 

Given the final matching feature vector of shape ```(batch_size * num_candidates, 16 * hidden_size)```, I pass it through a MLP, which has the following structure:
```
nn.Sequential(
            nn.Linear(16*hidden_size, 3*hidden_size),
            nn.ReLU(),
            nn.Dropout(p=dropout),
            nn.Linear(3*hidden_size, 1)
            )
```

I used Adam optimizer with learning rate of 0.001. The prediction layer is not learning, and as a result, my model produces logits that have the same value for each candidate. Have you encountered this problem before?

However, when I apply batch normalization before relu, the prediction layer starts learning and the model starts converging.

Any other tricks to accelerate and stablize the learning process?

Thanks,
Peixiang

**Comment by JasonForJoy** [comment]
@zhongpeixiang 
Sorry about that. I have not encountered this problem before.
In my original model, batch normalization is not applied and the training process is very stable. I did not experiment on whether it could improve the performance. 
One detail may be ignored that the learning rate is exponentially decayed by 0.96 every 5000 steps. You can have a try.
By the way, would you mind releasing the Pytorch version after debugging? Two versions of code may contribute to the NLP community. Thanks!

**Comment by zhongpeixiang** [comment]
Thank you for the guidance, I will try decaying learning rate in my experiments.

The DIM pytorch code will be released in this issue soon.
